pkg/logproto/compat.go

```diff
@@ -214,3 +214,91 @@ func init() {
 	jsoniter.RegisterTypeEncoderFunc("logproto.LegacySample", SampleJsoniterEncode, func(unsafe.Pointer) bool { return false })
 	jsoniter.RegisterTypeDecoderFunc("logproto.LegacySample", SampleJsoniterDecode)
 }
+
+// Combine unique values from multiple LabelResponses into a single LabelResponse.
+// TODO(jordanrushing): Should we sort labels lexicographically before returning?
+func MergeLabelResponses(responses []*LabelResponse) (*LabelResponse, error) {
+	if len(responses) == 0 {
+		return nil, fmt.Errorf("no label responses to merge")
+	} else if len(responses) == 1 {
+		return responses[0], nil
+	}
+
+	unique := map[string]struct{}{}
+
+	for _, r := range responses {
+		for _, v := range r.Values {
+			if _, ok := unique[v]; !ok {
+				unique[v] = struct{}{}
+			} else {
+				continue
+			}
+		}
+	}
+
+	result := &LabelResponse{Values: make([]string, 0, len(unique))}
+
+	for value := range unique {
+		result.Values = append(result.Values, value)
+	}
+
+	return result, nil
+}
+
+// Combine unique label sets from multiple SeriesResponse and return a single, sorted SeriesResponse.
+func MergeSeriesResponses(responses []*SeriesResponse) (*SeriesResponse, error) {
+	if len(responses) == 0 {
+		return nil, fmt.Errorf("no series responses to merge")
+	} else if len(responses) == 1 {
+		return responses[0], nil
+	}
+
+	unique := make(map[string]SeriesIdentifier)
+
+	for _, r := range responses {
+		for _, s := range r.Series {
+			key := LabelSet(s.Labels).String()
+			if _, ok := unique[key]; !ok {
+				unique[key] = s
+			} else {
+				continue
+			}
+		}
+	}
```

Comment:
Each tenant should have only unique series identifiers. If we then add the tenant label the merge series responses should be unique as well. Thus we don't require this bit.